---
title: "2021_StableGeneAnalysis"
author: "Eric Y."
date: "January 17, 2021"
output: html_document
---

```{r setup, include=FALSE}
setwd(dirname(rstudioapi::getActiveDocumentContext()$path))
library(ggplot2)
library(gghighlight)
library(ggrepel)
library(scales)
```

```{r Load in promoter/terminator data}
#ptl <- read.csv("Promoter_Terminator_length.csv")[,-1]
ptl <- read.csv("Promoter_Terminator_length_3percent.csv")[,-1]
ptl2 <- subset(ptl, UTR_promoter > UTR5_length & UTR_terminator > UTR3_length) #Ones where the neighboring gene doesn't intrude into promtoer region
Tissue4 <- Tissue3[Tissue3$AGI %in% ptl2$AGI,]

library("Biostrings")
upstream <- readDNAStringSet("Araport11_upstream_3000_translation_start_20160617")
downstream <- readDNAStringSet("Araport11_downstream_3000_translation_end_20160617")
ptl2$UTR_promoter_seq <- NA
ptl2$UTR_terminator_seq <- NA
ptl2$Promoter_seq <- NA

#For promtoers, regardless foward or reverse, the fasta file have 3000bp pointing towards the beginning of the gene, so to get the promoter + 5'UTR, just cound backwards from the last bp of the 3000bp number of bp according to UTR5+Promoter
#Conversly, for 3'UTR+Terminator always count from the first bp of the 3000bp.

#For each gene there are bound to be splice variants, and their UTR might not all start at the same location, for now default to ATXGXXXXX.1

for (n in 1:nrow(ptl2)){
  
  current_AGI <- ptl2[n,]$AGI
  UTR_promoter_length <- ptl2[n,]$UTR_promoter
  row_in_df <- which(stringr::str_detect(upstream@ranges@NAMES, pattern = paste0(current_AGI, ".1")))[1]
  if (is.na(row_in_df )){
    ptl2[n,]$UTR_promoter_seq <- NA
    ptl2[n,]$Promoter_seq <- NA
  } else {
  seq <- stringr::str_sub(upstream[[row_in_df]], -UTR_promoter_length , -1)
  ptl2[n,]$UTR_promoter_seq <- seq
  seq_pro <- stringr::str_sub(upstream[[row_in_df]], -UTR_promoter_length, -ptl2[n,]$UTR5_length-1)
  ptl2[n,]$Promoter_seq <- seq_pro
  }
  
  
  
  UTR_terminator_length <- ptl2[n,]$UTR_terminator
  row_in_df <- which(stringr::str_detect(downstream@ranges@NAMES, pattern = paste0(current_AGI, ".1")))[1]
  if (is.na(row_in_df)){
    ptl2[n,]$UTR_terminator_seq <- NA
  } else{
  seq <- stringr::str_sub(downstream[[row_in_df]], 1, UTR_terminator_length)
  ptl2[n,]$UTR_terminator_seq <- seq
  }
}

#Find internal cutsites for BbsI & BsaI
BbsI_f <- "GAAGAC"
BbsI_r <- "GTCTTC"
BsaI_f <- "GGTCTC"
BsaI_r <- "GAGACC"
CutSite <- c(BbsI_f, BbsI_r, BsaI_f, BsaI_r)

NoCutSite <- subset(ptl2, !(stringr::str_detect(UTR_promoter_seq, pattern = paste(CutSite, collapse = "|"))) & !(stringr::str_detect(UTR_terminator_seq, pattern = paste(CutSite, collapse = "|"))))

Tissue5 <- Tissue4[Tissue4$AGI %in% NoCutSite$AGI,]
```

```{r Plot genes with suitable promoter/terminators, and/or no cutsite}
ggplot(data = Tissue4, mapping = aes(x = geom_mean, y = CV)) +
  geom_point(aes(color = exclusive_core_type, shape = exclusive_core_type, fill = exclusive_core_type, alpha = 0.4), size = 4) +
  scale_shape_manual(values = c(21,22,23,24,25)) +
  theme_classic() +
  scale_x_continuous(trans='log2') +
  ggtitle(label = "Most stable genes by expression strength and core types") +
  xlab(label = "Geometric Mean (log2)") +
  ylab(label = "Coefficient of Variation") +
  guides(alpha = F) +
  labs(color = "Core Type", shape = "Core Type", fill = "Core Type")

ggplot(data = Tissue5, mapping = aes(x = geom_mean, y = CV)) +
  geom_point(aes(color = exclusive_core_type, shape = exclusive_core_type, fill = exclusive_core_type, alpha = 0.4), size = 4) +
  scale_shape_manual(values = c(21,22,23,24,25)) +
  theme_classic() +
  scale_x_continuous(trans='log2') +
  ggtitle(label = "Most stable genes by expression strength and core types") +
  xlab(label = "Geometric Mean (log2)") +
  ylab(label = "Coefficient of Variation") +
  guides(alpha = F) +
  labs(color = "Core Type", shape = "Core Type", fill = "Core Type")
```

```{r GuideSite}
#Always re-run, incase the last run was with different cutoff
file.create("Seq_Output.txt")
for (n in 1:nrow(NoCutSite)) {
  cat(c(paste0(">", NoCutSite[n,]$AGI), NoCutSite[n,]$Promoter_seq), file = "Seq_Output.txt", sep = "\n", append = TRUE)
}

#File uploaded to PlantRegMap & the result FIMO.txt is downloaded
#Always re-run, incase the last run was with different cutoff
Fimo <- read.table("fimo_003.txt", col.names = c("pattern name","Family","sequence name","start","stop","strand","score",	"p-value","q-value","matched sequence"))

guidesite_length <- 23

Output <- data.frame(
  AGI = character(),
  Region = double(),
  Overlaps = double()
)

for (n in unique(Fimo$sequence.name)){
  Max_length <- NoCutSite[which(NoCutSite$AGI == n),]$promoter_length
  sequence <- subset(Fimo, Fimo$sequence.name == n)
  
  Fimo_output <- data.frame(
  AGI = rep(n,Max_length-guidesite_length+1),
  Region = seq(1,Max_length-guidesite_length+1),
  Overlaps = rep(NA,Max_length-guidesite_length+1)
  )
  
  for (m in 1:nrow(Fimo_output)){
    guide_start <- Fimo_output$Region[m]
    guide_end <- guide_start + 22
    
    overlap <- subset(sequence, (sequence$start >= guide_start & sequence$start <= guide_end) | (sequence$stop >= guide_start & sequence$stop <= guide_end))
    
    Fimo_output$Overlaps[m] <- nrow(overlap)
    
  }
  
  Output <- rbind(Output, Fimo_output)
  
}

#Note the HIGHER the "region" in output, the closer it is to the TSS!!!
#Screen for two promtoers with two guide sites available, the dcas9 NOR gate paper had 67 bp between the two sites
GuideSite <- data.frame(
  AGI = character(),
  Pair1 = double(),
  Pair2 = double()
)

for (n in unique(Output$AGI)) {
  current_df <- subset(Output, AGI == n)
  if (nrow(current_df) >= 91) {
    for (m in nrow(current_df):91){
      if (current_df[m,]$Overlaps == 0) {
        
        pair2_df <- subset(current_df, Region <= m-90 & Overlaps == 0)
        
        if (nrow(pair2_df != 0)) {
          temp_output <- data.frame(
            AGI = n,
            Pair1 = m-(nrow(current_df)+1),
            Pair2 = max(pair2_df$Region)-(nrow(current_df)+1)
              #m-90-(nrow(current_df)+1)
          )
          GuideSite <- rbind(GuideSite, temp_output)
        }
      }
    }
  }
}

GuideSite_closetoTSS <- subset(GuideSite, Pair2 >= -500) #Increasing distance allowed doesn't grant more candidates from the Tissue3 list below.
unique(GuideSite_closetoTSS$AGI) #33 for 0.03 cutoff
```

```{r}
Tissue6 <- Tissue5[which(Tissue5$AGI %in% GuideSite_closetoTSS$AGI),]
ggplot(data = Tissue6, mapping = aes(x = geom_mean, y = CV)) +
  geom_point(aes(color = exclusive_core_type, shape = exclusive_core_type, fill = exclusive_core_type, alpha = 0.4), size = 4) +
  scale_shape_manual(values = c(21,22,23,24,25)) +
  theme_classic() +
  scale_x_continuous(trans='log2') +
  ggtitle(label = "Most stable genes by expression strength and core types") +
  xlab(label = "Geometric Mean (log2)") +
  ylab(label = "Coefficient of Variation") +
  guides(alpha = F) +
  labs(color = "Core Type", shape = "Core Type", fill = "Core Type")


library(ggrepel)
#No core type info
ggplot(data = Tissue7, mapping = aes(x = geom_mean, y = CV)) +
  geom_point(alpha = 0.4, size = 4, color = "steelblue") +
  #scale_shape_manual(values = c(21,22,23,24,25)) +
  theme_classic() +
  scale_x_continuous(trans='log2') +
  ggtitle(label = "Final 27 Candidates") +
  xlab(label = "Geometric Mean (log2)") +
  ylab(label = "Coefficient of Variation") +
  geom_text_repel(mapping = aes(label = AGI))+
  guides(alpha = F) +
  labs(color = "Core Type", shape = "Core Type", fill = "Core Type")

ggsave("27_candidates.png", width = 30, height = 20, units = "cm")
```


Remove the bottom 3 of expression strength with geom_mean cutoff of 430
```{r Final candidates}
Tissue7 <- subset(Tissue6, geom_mean >= 500) #Top 16 expression level

Tissue8 <- subset(Tissue7, StressCV <= 0.25)

Final_list <- merge(x = Tissue7, y = NoCutSite, by = "AGI", all.x = TRUE, all.y = FALSE)
write.csv(Final_list, "Candidate_Genes_fixed.csv") #Also wrote out Candidate_genes_fixed_003.csv
```

```{r}
ggplot(data = Tissue7, mapping = aes(x = geom_mean, y = CV)) +
  geom_point(aes(color = exclusive_core_type, shape = exclusive_core_type, fill = exclusive_core_type, alpha = 0.4), size = 4) +
  scale_shape_manual(values = c(21,22,23,24,25)) +
  theme_classic() +
  scale_x_continuous(trans='log2') +
  ggtitle(label = "Final 27 candidates by expression strength and core types") +
  xlab(label = "Geometric Mean (log2)") +
  ylab(label = "Coefficient of Variation") +
  guides(alpha = F) +
  labs(color = "Core Type", shape = "Core Type", fill = "Core Type")

```

Final list contains 20 candidates:
Geom_mean range from 537~6683
7/20 also has Stress CV <=0.25 (i.e. if I started off using both dataset <=0.25, I would end up with 7)

```{r Figure2 Inset Plot log2}
largeplot <- ggplot(data = Tissue6, mapping = aes(x = geom_mean, y = CV)) +
  geom_point(alpha = 0.4, size = 4) +
  theme_bw() +
  scale_x_continuous(trans='log2', limits = c(256, 8192), breaks=scales::trans_breaks("log2", function(x) 2^x)) +
  scale_y_continuous(breaks=seq(0.12, 0.26, 0.02), limits = c(0.12,0.26)) +
  #annotation_logticks(sides="b", base = 10) +
  xlab(label = "Geometric Mean (log2)") +
  ylab(label = "Coefficient of Variation") +
  theme(axis.title =element_text(family="sans"))
  

smallplot <- ggplot(data = Tissue3, mapping = aes(x = geom_mean, y = CV)) +
  geom_point(alpha = 0.4) +
  theme_bw() +
  scale_x_continuous(trans='log2', breaks=c(256,1024,8192)) +
  scale_y_continuous(breaks=seq(0.12, 0.26, 0.07), limits = c(0.12,0.26)) +
  #ggtitle("lowest 3% CV")+
  theme(axis.title.x=element_blank(),
      axis.title.y=element_blank()#,
      #plot.title=element_text(size = 12, vjust = -40, hjust = 0.12, margin=margin(t=10,b=-20), family = "sans")
      )

largeplot + annotation_custom(ggplotGrob(smallplot), xmin = 10, xmax = 12.9, ymin = 0.142, ymax = 0.222)
```


```{r}
stable_list <- read.csv("Cheng_Klep_Kudo_Me.csv")
sum(unique(stable_list$My_analysis) %in% unique(stable_list$Map3))
sum(unique(stable_list$My_analysis) %in% unique(stable_list$Map25)) -1 #remove NA
```
201/202 of my top 2% overlaps with Klep CV 3 cut off.
168/202 of my top 2% overlaps with Klep CV 2.5 cut off.

```{r Manually process fimo for benchling}
file_path <- "C:/Users/ericy/OneDrive - UW/Nemhuaser Lab/Pest-induced Defense/PromoterScreen/2021_StableGeneAnalysis/Fimo_Output/"

file_name <- "AT1G71900"

Fimo <- read.table(paste0(file_path,file_name, ".txt"), col.names = c("pattern name","Family","sequence name","start","stop","strand","score",	"p-value","q-value","matched sequence"))

Fimo <- Fimo[,c(2,10)]
colnames(Fimo) <- c("name", "feature")
write.csv(Fimo, paste0(file_path, file_name, ".csv"),row.names = FALSE)
```


```{r}
#Display the top five stable genes
MoR_Tissue <- read.csv("MoR_TissueOnly.csv")
Tissue <- read.csv("MoR_Stats_Tissue.csv")[,-1]
Tissue_ranked <- Tissue
Tissue_ranked$rank <- NA
order.scores<-order(Tissue_ranked$CV)
Tissue_ranked$rank[order.scores] <- 1:nrow(Tissue_ranked)
Top5 <- subset(Tissue_ranked, rank <= 5)
Top5_reads <- subset(MoR_Tissue, X %in% Top5$AGI)
Top5_reads2 <- t(Top5_reads)
colnames(Top5_reads2) <- Top5_reads2[1,]
Top5_reads2 <- Top5_reads2[-1,]
Top5_reads3 <- reshape2::melt(data = Top5_reads2)
Top5_reads4 <- as.data.frame(Top5_reads3)
Top5_reads4$value <- as.double(Top5_reads4$value)
Top5_reads4$Var2 <- factor(Top5_reads4$Var2, levels = c("AT5G06140", "AT1G54080", "AT5G58440", "AT2G35330", "AT5G18230"))

ggplot(Top5_reads4, mapping = aes(x=Var2, y= value)) +
  geom_violin() +
  geom_jitter(width = 0.2, alpha = 0.4) +
  theme_classic() +
  ggtitle("Top Five Stable Genes by Coefficient of Variation")+
  xlab("Gene Name (ranked from left to right)")+
  ylab("Normalized Read Counts")
```

```{r}
#Display the top five stable genes
MoR_Tissue <- read.csv("MoR_TissueOnly.csv")
Tissue <- read.csv("MoR_Stats_Tissue.csv")[,-1]
Tissue_ranked <- Tissue
Tissue_ranked$rank <- NA
order.scores<-order(Tissue_ranked$CV)
Tissue_ranked$rank[order.scores] <- 1:nrow(Tissue_ranked)


#0.13, 0.2, 0.25, 0.3, 0.35, 0.4, 0.6, 0.8, 1.0
candidates <- subset(Tissue_ranked, AGI %in% c("AT5G06140", "AT2G47210", "AT5G12210", "AT4G29940", "AT3G21865", "AT5G03040", "AT3G61440", "AT5G61670", "AT3G18500"))
candidates_reads <- subset(MoR_Tissue, X %in% candidates$AGI)
Top5_reads2 <- t(candidates_reads)
colnames(Top5_reads2) <- Top5_reads2[1,]
Top5_reads2 <- Top5_reads2[-1,]
Top5_reads3 <- reshape2::melt(data = Top5_reads2)
Top5_reads4 <- as.data.frame(Top5_reads3)
Top5_reads4$value <- as.double(Top5_reads4$value)
Top5_reads4$Var2 <- factor(Top5_reads4$Var2, levels = c("AT5G06140", "AT2G47210", "AT5G12210", "AT4G29940", "AT3G21865", "AT5G03040", "AT3G61440", "AT5G61670", "AT3G18500"))

ggplot(Top5_reads4, mapping = aes(x=Var2, y= value)) +
  geom_boxplot() +
  theme_classic() +
  scale_y_continuous(trans = 'log2') +
  ggtitle("Genes ranked by CV")+
  xlab("Gene Name (ranked from left to right)")+
  ylab("Normalized Read Counts")
```

```{r Stress vs Map}
Tissue <- read.csv("MoR_Stats_Tissue.csv")[,-1] #10096
Stress <- read.csv("MoR_Stats_StressOnly.csv")[,-1] #16634

colnames(Tissue)[colnames(Tissue) == "CV"] <- "CV_Tissue"
colnames(Tissue)[colnames(Tissue) == "geom_mean"] <- "Mean_Tissue"
colnames(Stress)[colnames(Stress) == "CV"] <- "CV_Stress"
colnames(Stress)[colnames(Stress) == "geom_mean"] <- "Mean_Stress"

sum((Tissue$AGI %in% Stress$AGI))

Tissue2 <- subset(Tissue, AGI %in% Stress$AGI)
Stress2 <- subset(Stress, AGI %in% Tissue$AGI)

cutoff <- 0.03
Tissue3 <- subset(Tissue2, CV <= quantile(Tissue2$CV, cutoff))
Stress3 <- subset(Stress2, CV <= quantile(Stress2$CV, cutoff))
sum((Tissue3$AGI %in% Stress3$AGI))

#plot
cutoff <- 0.03
Tissue2 <- subset(Tissue, CV_Tissue <= quantile(Tissue$CV_Tissue, cutoff))

Combined <- merge(x = Tissue2, y= Stress, by = "AGI", all.x = F, all.y = F)

ggplot(data = Combined, mapping = aes(x = CV_Tissue, y = CV_Stress)) +
  geom_point(mapping = aes(size = Mean_Tissue), alpha = 0.4)+
  ggtitle("Top 3% Stable Promoters, Tissue CV vs Stress CV") +
  xlab("Tissue CV") +
  ylab("Stress CV (log)") +
  geom_abline(slope = 1, intercept = 0) +
  coord_trans(y="log") +
  guides(size = guide_legend(title = "Mean Expression \n(normalized read count)"))+
  theme_classic()+
  theme(legend.title = element_text(size = 9))
  
  

#plot w/o cutoff
Combined2 <- merge(x = Tissue, y = Stress, by = "AGI", all.x = F, all.y = F)
ggplot(data = Combined2, mapping = aes(x = CV_Tissue, y = CV_Stress)) +
  geom_point(mapping = aes(size = Mean_Tissue, alpha = 0.4)) +
  scale_x_continuous(trans='log2') +
  scale_y_continuous(trans='log2')

```

```{r CV vs CV}
candidate <- read.csv("Candidate_Genes_fixed_003.csv")

ggplot(data = candidate, mapping = aes(x = CV, y = StressCV)) +
  geom_point(mapping = aes(size = geom_mean, alpha = 0.4)) +
  scale_x_continuous(breaks = seq(0, 0.3, 0.05), 
                     labels = seq(0, 0.3, 0.05))+ 
  ggtitle("Final Candidates Stress CV vs Tissue CV") +
  xlab("Tissue CV") +
  ylab("Stress CV")
```

```{r Intersection with TPR1-targets}
TPR1<-read.table("C:/Users/ericy/OneDrive - UW/Nemhuaser Lab/Griebel_TPR1_Github/R_metagene_TPR1/TPR1_targets.txt")
Tissue7$AGI %in% TPR1$V1
```


```{r All promoters}
#run first bit of code to get genes annotated, stress data included since it doesn't drop any tissue data.
#Multiple ATGs do not have annoated .1 variant. Those have NA entries and dropped subsequently
#Upstream and Downstream annoated file contains situation where a single ATG.1 variant will be mappted to multiple entires, the first one listed is picked. Again, if .1 variant is not found, NA is added then subsequently dropped.
ptl <- read.csv("Promoter_Terminator_length_ALL.csv")[,-1]
ptl2 <- subset(ptl, UTR_promoter > UTR5_length & UTR_terminator > UTR3_length) #Ones where the neighboring gene doesn't intrude into promtoer region
ptl2 <- ptl2[is.finite(rowSums(ptl2[,c(-1,-2)])),] #No Inf in there 6951 obs
Tissue4 <- Tissue3[Tissue3$AGI %in% ptl2$AGI,]
#write.csv(ptl2, "Sequence_annotated_ptl2_ALL.csv")
ptl3 <- ptl2[complete.cases(ptl2), ] #6913 without NAs
NoCutSite <- ptl3 #Just to make the code compatable without adding lines

Fimo <- read.table("fimo_003.txt", col.names = c("pattern name","Family","sequence name","start","stop","strand","score",	"p-value","q-value","matched sequence"))


Fimo <- read.table("fimo_003.txt", col.names = c("pattern name","Family","sequence name","start","stop","strand","score",	"p-value","q-value","matched sequence"))
Fimo <- Fimo[0,]

for (n in list.files("C:/Users/ericy/OneDrive - UW/Nemhuaser Lab/Pest-induced Defense/PromoterScreen/04222021_Motif_Analysis/Fimo", full.names = TRUE)) {
  fimo <- read.table(n, col.names = c("pattern name","Family","sequence name","start","stop","strand","score",	"p-value","q-value","matched sequence"))
  Fimo <- rbind(Fimo, fimo)
}

#write.csv(Fimo, "All_Fimo.csv")

Fimo2 <- Fimo[,c(1,2,3,4,5)]
Fimo2$From_Tss <- NA
for (n in unique(Fimo2$sequence.name)) {
  Fimo2$From_Tss[which(Fimo2$sequence.name == n)] <- (as.double(ptl3$promoter_length[which(ptl3$AGI == n)]) - Fimo2$start[which(Fimo2$sequence.name == n)])
}

Fimo3 <- Fimo2[c(1,2,3,6)]

Fimo4 <- merge(Fimo3, Tissue3, by.x = "sequence.name", by.y = "AGI", all.x = TRUE, all.y = FALSE)

write.csv(Fimo4, "Motif_loc_coretype.csv")

Fimo <- read.csv("C:/Users/ericy/OneDrive - UW/Nemhuaser Lab/Pest-induced Defense/PromoterScreen/04222021_Motif_Analysis/Fimo/All_Fimo.csv")[,-1]

Tissue4$Motif_Count <- NA
for (n in Tissue4$AGI){
  current_df <- subset(Fimo, sequence.name == n)
  Tissue4$Motif_Count[which(Tissue4$AGI == n)] <- nrow(current_df)
}

prom_length <- ptl3[,c(1,15)]

Tissue4 <- merge(Tissue4, ptl3[,c(1,15)], by = "AGI", all.x = TRUE, all.y = FALSE)

Tissue5 <- merge(Tissue4, temp[,c(2,12)], by = "AGI", all.x = TRUE, all.y = FALSE)

write.csv(Tissue5, "Exploratory_Analysis.csv")


Fimo4 <- read.csv("C:/Users/ericy/OneDrive - UW/Nemhuaser Lab/Pest-induced Defense/PromoterScreen/04222021_Motif_Analysis/Motif_loc_coretype.csv")[,-1]

Fimo5 <- subset(Fimo4, From_Tss <= 500)
Tissue4$Motif_Count <- NA
for (n in Tissue4$AGI){
  current_df <- subset(Fimo5, sequence.name == n)
  Tissue4$Motif_Count[which(Tissue4$AGI == n)] <- nrow(current_df)
}

write.csv(Tissue4, "All_promoters_Motif_within500.csv")
```


```{r Annotating promoter/terminator length, eval=FALSE}
#First summarize gtf by finding the earilest start position and last end position of each ATG number
gtf <- rtracklayer::import('Araport11_GTF_201606.gtf') #For some reason all the bp are shifted right 1bp, need to adjust
gtf_df=as.data.frame(gtf)
gtf_df$gene_id <- substr(gtf_df$gene_id, 1, 9)
gtf_df_gene <- subset(gtf_df, type == "gene") #Used to determine gene neighbors, all splice variants included here

#!!!!!!!!!!!!!!!!Be very careful that "start" and "end" means different things depending on the strand, but start is always to the left of end, regardless of strand!!!!!!!!!!

output_df <- data.frame(
    AGI = character(),
    strand = character(),
    UTR5_start = double(),
    UTR5_end = double(),
    UTR5_length = double(),
    UTR3_start = double(),
    UTR3_end = double(),
    UTR3_length = double(),
    right_gene_start = double(),
    UTR_terminator = double(),
    left_gene_end = double(),
    UTR_promoter = double(),
    promoter_start = double(),
    promoter_end = double(),
    promoter_length = double()
)


#Add a Gene_start and Gene_end to find intrution of neighbors

    #UTR5_start = 0
    #UTR5_end = 0
    #UTR5_length = 0
    #UTR3_start = 0
    #UTR3_end = 0
    #UTR3_length = 0
    #right_gene_start = 0
    #UTR_terminator = 0
    #left_gene_end = 0
    #UTR_promoter = 0

for (m in unique(Tissue3$AGI)){
  goi <- subset(gtf_df, gene_id == m)
  goi_gene.1 <- subset(goi, transcript_id == paste0(m, ".1"))
  if (nrow(goi_gene.1) != 0) {
  #If the gene is on + strand:
  if (goi_gene.1[1,]$strand == "+"){
    UTR5_df <- subset(goi, type == "five_prime_UTR" & stringr::str_detect(goi$transcript_id, pattern = paste0(m, ".1")))
    
    if (nrow(UTR5_df) == 0) {
      UTR5_start <- min(goi_gene.1$start) -1
      UTR5_end <- min(goi_gene.1$start) -1
      UTR5_length <- UTR5_end - UTR5_start
    }
    
    if (nrow(UTR5_df) != 0) {
    UTR5_start <- min(UTR5_df$start) -1
    UTR5_end <- max(UTR5_df$end) -1
    UTR5_length <- UTR5_end - UTR5_start + 1
    }
    
    UTR3_df <- subset(goi, type == "three_prime_UTR" & stringr::str_detect(goi$transcript_id, pattern = paste0(m, ".1")))
    
    if (nrow(UTR3_df) == 0) {
      UTR3_start <- max(goi_gene.1$end) -1
      UTR3_end <- max(goi_gene.1$end) -1
      UTR3_length <- UTR3_end - UTR3_start
    }
    
    if (nrow(UTR3_df) != 0) {
    UTR3_start <- min(UTR3_df$start) -1 #need
    UTR3_end <- max(UTR3_df$end) -1 #need
    UTR3_length <- UTR3_end - UTR3_start + 1 #need
    }
    
    #gene to the right:
    right_gene_AGI <- unique(gtf_df_gene$gene_id)[which(unique(gtf_df_gene$gene_id) == m) + 1]
    right_gene <- subset(gtf_df, gene_id == right_gene_AGI & type == "gene")
    right_gene_start <- min(right_gene$start) -1 #need
    UTR_terminator <- right_gene_start - UTR3_start - 1 #need
    
    #gene to the left:
    left_gene_AGI <- unique(gtf_df_gene$gene_id)[which(unique(gtf_df_gene$gene_id) == m) - 1]
    left_gene <- subset(gtf_df, gene_id == left_gene_AGI & type == "gene")
    left_gene_end <- max(left_gene$end) -1 #need
    UTR_promoter <- UTR5_end - left_gene_end  #need
    
    if (UTR_promoter > UTR5_length) {
        promoter_start <- left_gene_end + 1
        promoter_end <- UTR5_start -1
      
    } else {
      promoter_start <- left_gene_end + 1
      promoter_end <- UTR5_start
    }
    promoter_length <- promoter_end - promoter_start + 1

    if (nrow(UTR5_df) == 0) {UTR_promoter <- promoter_length}
  }
  
  #If the gene is on - strand:
  if (goi_gene.1[1,]$strand == "-"){
    UTR5_df <- subset(goi, type == "five_prime_UTR" & stringr::str_detect(goi$transcript_id, pattern = paste0(m, ".1")))
    
    if (nrow(UTR5_df) == 0) {
      UTR5_start <- max(goi_gene.1$end) -1
      UTR5_end <- max(goi_gene.1$end) -1
      UTR5_length <- UTR5_end - UTR5_start
    }
    
    if (nrow(UTR5_df) != 0) {
    UTR5_start <- min(UTR5_df$start) -1 #need
    UTR5_end <- max(UTR5_df$end) -1 #need
    UTR5_length <- UTR5_end - UTR5_start + 1 #need
    }
    
    
    UTR3_df <- subset(goi, type == "three_prime_UTR" & stringr::str_detect(goi$transcript_id, pattern = paste0(m, ".1")))
    
    if (nrow(UTR3_df) == 0) {
      UTR3_start <- min(goi_gene.1$start)  -1
      UTR3_end <- min(goi_gene.1$start) -1
      UTR3_length <- UTR3_end - UTR3_start
    }
    
    if (nrow(UTR3_df) != 0) {
    UTR3_start <- min(UTR3_df$start) -1 #need
    UTR3_end <- max(UTR3_df$end) -1 #need
    UTR3_length <- UTR3_end - UTR3_start + 1 #need
    }
    
    #gene to the right:
    right_gene_AGI <- unique(gtf_df_gene$gene_id)[which(unique(gtf_df_gene$gene_id) == m) + 1]
    right_gene <- subset(gtf_df, gene_id == right_gene_AGI & type == "gene")
    right_gene_start <- min(right_gene$start) -1 #need
    UTR_promoter <- right_gene_start - UTR5_start #need
    
    #gene to the left:
    left_gene_AGI <- unique(gtf_df_gene$gene_id)[which(unique(gtf_df_gene$gene_id) == m) - 1]
    left_gene <- subset(gtf_df, gene_id == left_gene_AGI & type == "gene")
    left_gene_end <- max(left_gene$end) -1 #need
    UTR_terminator <- UTR3_end - left_gene_end - 1 #need
    
    if (UTR_promoter > UTR5_length) {
        promoter_start <- UTR5_end + 1
        promoter_end <- right_gene_start -1
      
    } else{
      promoter_start <- UTR5_start
      promoter_end <- right_gene_start -1
    }
    
    promoter_length <- promoter_end - promoter_start + 1
    if (nrow(UTR5_df) == 0) {UTR_promoter <- promoter_length}
  }

  current_output <- data.frame(
    AGI = m,
    strand = goi_gene.1[1,]$strand,
    UTR5_start = UTR5_start,
    UTR5_end = UTR5_end,
    UTR5_length = UTR5_length,
    UTR3_start = UTR3_start,
    UTR3_end = UTR3_end,
    UTR3_length = UTR3_length,
    right_gene_start = right_gene_start,
    UTR_terminator = UTR_terminator,
    left_gene_end = left_gene_end,
    UTR_promoter = UTR_promoter,
    promoter_start = promoter_start,
    promoter_end = promoter_end,
    promoter_length = promoter_length
  )
  } else { #If there is no .1 found
    current_output <- data.frame(
    AGI = m,
    strand = goi[1,]$strand,
    UTR5_start = NA,
    UTR5_end = NA,
    UTR5_length = NA,
    UTR3_start = NA,
    UTR3_end = NA,
    UTR3_length = NA,
    right_gene_start = NA,
    UTR_terminator = NA,
    left_gene_end = NA,
    UTR_promoter = NA,
    promoter_start = NA,
    promoter_end = NA,
    promoter_length = NA)
  }
  
  output_df <- rbind(output_df, current_output)
}
write.csv(output_df, "Promoter_length_no2bpRestriction.csv")
```

```{r Annotate gene symbol}
library("Biostrings")
CDS <- readDNAStringSet("Araport11_cds_20160703_representative_gene_model")

Output <- data.frame(
  AGI = character(),
  Symbol = character(),
  Full_Name = character()
)

for (n in 1:length(CDS@ranges)) {
  current_string <- strsplit(CDS@ranges[n]@NAMES, "\\|")
  current_AGI <- current_string[[1]][1]
  current_symbol <- current_string[[1]][2]
  current_name <- current_string[[1]][3]

  current_AGI <- strsplit(current_AGI, "\\.")[[1]][1]
  current_symbol <- strsplit(current_symbol, "\\:")[[1]][2]
  
  new_line <- data.frame(
    AGI = current_AGI,
    Symbol = current_symbol,
    Full_Name = current_name
  )
  
  Output <- rbind(Output, new_line)
}

write.csv(Output, "AGItoSymbol.csv")
```

```{r}
expanalysis <- read.csv("Exploratory_Analysis.csv")[,-1]
output <- merge(expanalysis, Output, by = "AGI", all.x = TRUE, all.y = FALSE)

write.csv(output, "Exploratory_Analysis.csv")
```

To Do:
Use 5% top stable, rank with top 202 stable by stress then proceed
****append promoter/terminator length
****see how many TPR are in the top 2% stable genes

Analyze difference in the stable subset to total dataset in additional genic/antisense/ etc.
Compare my tissue stable genes to Klepikova map dataset
Compare my final list to Kelpikova Genes table in both map & map+Stress
Compare top tissue + ranking vs all normalize together